### PR TITLE
Wip/hughsie/flatpak

### PR DESCRIPTION
**Pull request by hughsie**
This allows us to compile without PolicyKit, and thus we only install the fwupdtool binary and not the daemon (which won't work in the flatpak model).

### CODE REVIEW

File: meson_options.txt [hunk 1]
```diff
@@ -1,4 +1,6 @@
 option('bootdir', type : 'string', value : '/boot/efi', description : 'Directory for EFI system partition')
+option('polkit', type : 'boolean', value : true, description : 'enable PolicyKit support')
```
Comment (superm1) [inline meson_options.txt]: Since this is affecting the installation of the daemon and not just the installation of the rules I think this string should be renamed to "standalone".  It can impact anything it needs to then that doesn't already have it's own string.

File: meson_options.txt [hunk 2]
```diff
@@ -1,4 +1,6 @@
...
+option('udevrules', type : 'boolean', value : true, description : 'enable UDev rules')
```
Comment (superm1) [inline meson_options.txt]: Also this option then can be avoided as standalone can cover both.

File: src/meson.build [hunk 3]
```diff
@@ -183,6 +180,7 @@ if get_option('man')
   )
 endif
 
+if get_option('polkit')
```
Comment (superm1) [inline src/meson.build]: Pretty sure you'll need the same wrapper around fwupdmgr's executable too.

Comment (superm1) [reply src/meson.build]: I think it's going to be inviting bugs, particularly further down the road when fwupd grows new APIs that require upgrading both.

File: plugins/meson.build [hunk 4]
```diff
@@ -18,7 +18,7 @@ if get_option('plugin_amt')
 subdir('amt')
 endif
 
-if get_option('plugin_thunderbolt') and udev.found()
+if get_option('plugin_thunderbolt')
```
Comment (superm1) [inline plugins/meson.build]: You should split this into a separate commit explaining that udev isn't in the build env but a runtime dep etc.

File: meson_options.txt (hunk 1 above)
Comment (hughsie) [reply meson_options.txt]: Usually configure options are default on, and let you disable some action. I think a -Ddaemon=false and then checking for an option makes more sense than a -Dstandalone and then checking for the negative of the option. I've fixed up the commit to add a daemon=false mode -- does that look sane?

File: plugins/meson.build (hunk 4 above)
Comment (hughsie) [reply plugins/meson.build]: Right, I'll push that chunk right to master. Thanks.

File: meson_options.txt (hunk 1 above)
Comment (superm1) [reply meson_options.txt]: I think that logic makes sense, yeah.

File: src/meson.build (hunk 3 above)
Comment (hughsie) [reply src/meson.build]: True. I'll ifdef it.

File: data/meson.build [hunk 5]
```diff
@@ -24,9 +24,16 @@ install_data(['metadata.xml'],
   install_dir : join_paths(datadir, 'fwupd', 'remotes.d', 'fwupd')
 )
 
-install_data(['90-fwupd-devices.rules'],
-  install_dir : join_paths(udevdir, 'rules.d')
-)
+if get_option('daemon')
+  udevdir = get_option('udevdir')
+  if udevdir == ''
+    udev = dependency('udev')
```
Comment (superm1) [inline data/meson.build]: I'd prefer to keep all dependency selection logic at the top level meson.build so it's easy to find what options are introducing new dependencies.  Would you mind to move conditional udev dependency there?

File: meson.build [hunk 6]
```diff
@@ -134,9 +134,11 @@ add_project_arguments('-D_GNU_SOURCE', language : 'c')
...
 gmodule = dependency('gmodule-2.0')
 giounix = dependency('gio-unix-2.0', version : '>= 2.45.8')
-polkit = dependency('polkit-gobject-1', version : '>= 0.103')
-if polkit.version().version_compare('>= 0.114')
-  conf.set('HAVE_POLKIT_0_114', '1')
+if get_option('daemon')
+  polkit = dependency('polkit-gobject-1', version : '>= 0.103')
+  if polkit.version().version_compare('>= 0.114')
+    conf.set('HAVE_POLKIT_0_114', '1')
+  endif
```
Comment (superm1) [inline meson.build]: As in underneath this get_option ('daemon')

**Comment by superm1** [comment]
My other thought in the selection of the string for this option is I'd like the snap runtime env strings to ideally turn on when those are eventually merged from this same option.  Do you think that makes sense or those should stay as their own separate option?  

I suspect that with the changes in this PR I might actually be able to fix the current problems that was making snap build fail on 16.04 (policy kit).

**Comment by superm1** [comment]
> Is the snap going to build the daemon or just fwupdtool? I guess if both then I'd need to see what the options look like.

I think initially I want to do just fwupdtool.  Doing the daemon too I think can be a stretch goal for later and we can change options if need be then.
